### PR TITLE
Test gas consumption of EM.run()

### PR DESCRIPTION
**Pull request by maurelian**
## Description

Adds a utility class for gas measurement, and a test case that expects the EM's gas use to fall within a given range.

### CODE REVIEW

File: test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
+import { expect } from '../../../setup'
+import { deployContractCode } from '../../../helpers/utils'
+
+/* External Imports */
+import { ethers } from 'hardhat'
+import { Contract, ContractFactory, Signer } from 'ethers'
+import { smoddit, smockit, MockContract } from '@eth-optimism/smock'
+import _ from 'lodash'
```
Comment (kfichter) [inline test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts:8]: Worth double checking that all of these imports are being used.

File: test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      
+      let benchmark:number = 226_516
+      expect(gasCost).to.be.lte(benchmark)
+      expect(gasCost).to.be.gte(
+        benchmark - 1_000,
+        "Gas cost has significantly decreased, consider updating the benchmark to reflect the change"
+      )
+    })
+  })
+})
```
Comment (kfichter) [inline test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts:130]: Nit to add newlines at the end of any files in this PR that don't have them.

File: test/helpers/utils/custom-deployer.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { Signer } from 'ethers'
+import { toHexString }  from '../../../src/utils' 
+
+export const deployContractCode = async (code: string, signer: Signer, gasLimit: number): Promise<string> => {
```
Comment (kfichter) [inline test/helpers/utils/custom-deployer.ts:4]: Would be good to eventually put this into `core-utils` but fine to copy/paste here instead for now.

File: test/helpers/gas/gas.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { ethers } from 'hardhat'
+import { Contract, Signer } from 'ethers'
+
+export class GasMeasurement {
```
Comment (kfichter) [inline test/helpers/gas/gas.ts:4]: Can this be used to measure the gas usage of any function? I assume so, just want to double check.

**Review by kfichter — APPROVED** [review]
LGTM, asked for a few small nits but nothing blocking.

File: test/helpers/gas/gas.ts (hunk 4 above)
Comment (maurelian) [reply test/helpers/gas/gas.ts:4]: That's the intention, and I think it achieves that.

**Review by kfichter — APPROVED** [review]
LGTM

**Pull request merged**